File: backend/app/core/parser/ruby.py
import re
from pathlib import Path
from typing import List

from app.api.models import ImportInfo


class RubyParser:
    """Parser for Ruby files."""

    def get_supported_extensions(self) -> List[str]:
        """Get supported file extensions."""
        return [".rb"]

    def parse_file(self, file_path: Path) -> List[ImportInfo]:
        """Parse a Ruby file and extract require statements.

        Args:
            file_path: Path to the file

        Returns:
            List of ImportInfo objects
        """
        imports = []
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()

            # Handle optional parentheses: require("foo") or require "foo"
            patterns = [
                (r"""require\s*\(?\s*['"]([^'"]+)['"]""", "require"),
                (r"""require_relative\s*\(?\s*['"]([^'"]+)['"]""", "require_relative"),
                (r"""load\s*\(?\s*['"]([^'"]+)['"]""", "load"),
            ]

            for i, line in enumerate(content.split("\n"), 1):
                line = line.strip()
                if line.startswith("#"):
                    continue

                for pattern, import_type in patterns:
                    matches = re.finditer(pattern, line)
                    for match in matches:
                        imports.append(
                            ImportInfo(
                                source_file=str(file_path),
                                imported_module=match.group(1),
                                import_type=import_type,
                                line_number=i,
                            )
                        )
        except Exception:
            pass
        return imports

File: extractor/extract/ruby.go
package extract

import (
	"regexp"
	"strings"
)

var (
	rubyRequireRegex         = regexp.MustCompile(`require\s*\(?\s*['"]([^'"]+)['"]`)
	rubyRequireRelativeRegex = regexp.MustCompile(`require_relative\s*\(?\s*['"]([^'"]+)['"]`)
	rubyLoadRegex            = regexp.MustCompile(`load\s*\(?\s*['"]([^'"]+)['"]`)
)

// IsRubyFile checks if a file is a Ruby file based on extension.
func IsRubyFile(path string) bool {
	return strings.HasSuffix(strings.ToLower(path), ".rb")
}

// ExtractRuby extracts imports from a Ruby source file.
func ExtractRuby(content string, filePath string) []Import {
	var imports []Import
	lines := strings.Split(content, "\n")

	patterns := map[*regexp.Regexp]string{
		rubyRequireRegex:         "require",
		rubyRequireRelativeRegex: "require_relative",
		rubyLoadRegex:            "load",
	}

	for i, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "#") {
			continue
		}

		for re, importType := range patterns {
			matches := re.FindAllStringSubmatch(line, -1)
			for _, match := range matches {
				if len(match) > 1 {
					imports = append(imports, Import{
						SourceFile:     filePath,
						ImportedModule: match[1],
						ImportType:     importType,
						LineNumber:     i + 1,
					})
				}
			}
		}
	}
	return imports
}